### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by kaykay38**
Collaborated with Kiana Norton @NyahHuh

### CODE REVIEW

**Comment by kaykay38** [comment]
## Assignment

- There is an existing `BaseLogger` class. It needs an **auto property** to hold the class name. This property should be set in the `LogFactory` using an **object initializer**. ✔
- Create a `FileLogger` that derives from `BaseLogger`. It should take in a path to a file to write the log message to. When its `Log` method is called, it should **append** messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The `LogFactory` should be updated with a new method `ConfigureFileLogger`. This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method. ✔
- If the file logger has not be configured in the `LogFactory`, its `CreateLogger` method should return `null`. ✔
- Inside of `BaseLoggerMixins` implement **extension methods** on `BaseLogger` for
  - `Error`, ✔
  - `Warning`, ✔
  - `Information`, and ✔
  - `Debug`. ✔
  Each of these methods should take in a `string` for the message, as well as a **parameter array** of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the `BaseLogger.Log` method, by automatically supplying the appropriate `LogLevel`. These methods should throw an exception if the `BaseLogger` parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ✔
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
- Set `LangVersion` to 9.0 ✔
- Turn on Nullability (`Nullable`) ✔
- **All of the above should be unit tested.** ❌

## Extra Credit

- Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses `System.Console` or `System.Diagnostics.Trace` ❌

**Comment by NyahHuh** [comment]
Assignment
•	There is an existing BaseLogger class. It needs an auto property to hold the class name. This property should be set in the LogFactory using an object initializer. ✔
•	Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
o	The current date/time ✔
o	The name of the class that created the logger ✔
o	The log level ✔
o	The message ✔
o	The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
•	The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
•	If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
•	Inside of BaseLoggerMixins implement extension methods on BaseLogger for
o	Error, ✔
o	Warning, ✔
o	Information, and ✔
o	Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
•	Use the nameof() operator when identifying the class name to the logger ✔
•	Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ❌ (Unsure, know we found one of the TreatWarningsAsErrors in the Project Properties to All)
•	Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
•	Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
•	Set LangVersion to 9.0 ✔
•	Turn on Nullability (Nullable) ✔
•	All of the above should be unit tested. ❌ We need to finish the unit testing to properly run and test everything. 
Extra Credit
•	Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses System.Console or System.Diagnostics.Trace ❌


Will go back in and finish the unit tests if is possible.

File: Logger/FileLogger.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        private string? _filePath;
+        public string? FilePath
+        {
+            get
+            {
+                return _filePath;
+            }
+            set => _filePath = value;
+        }
+        public FileLogger(string filePath)
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:17]:
```suggestion
        public FileLogger(string? filePath)
```
This should be nullable if you're throwing a null exception below.
Setting it to nullable will also allow you to remove the `!` operator in your unit tests.

File: Logger.Tests/FileLoggerTests.cs [hunk 2]
```diff
@@ -1,10 +1,71 @@
...
+{
+    [TestClass]
+    public class FileLoggerTests
+    {
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void FileLogger_NullFile_ThrowsArgumentNullException()
+        {
+            // Act
+            FileLogger? fileLogger = new FileLogger(null!);
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:16]:
```suggestion
            FileLogger? fileLogger = new FileLogger(null);
```
`!` can be removed if you set the file name parameter to be nullable.

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 3]
```diff
@@ -1,48 +1,54 @@
...
-        }
-    }
-}
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+using System.Collections.Generic;
+
+namespace Logger.Tests
+{
+    [TestClass]
```
Comment (mmwoodfo) [inline Logger.Tests/BaseLoggerMixinsTests.cs:7]: Missing tests for warning, debug and information.

**Comment by mmwoodfo** [comment]
There are 35 changed files and an apparent merge conflict. Did you use git to clone this repo & branch the assignment?

**Pull request closed without merging**